### ISSUE TITLE
False positive CS8618 null reference type warning when using required members and [SetsRequiredMembers] in derived class

### ISSUE THREAD

**Issue by SapiensAnatis** [dotnet/roslyn#74423]
**Version Used**: 

`Compiler version: '4.10.0-3.24314.14 (259e82e9)'. Language version: 12.0.`, but also reproduces on `main (11 Jun 2024)` on Sharplab.

**Steps to Reproduce**:

1. Compile the following code ([sharplab](https://sharplab.io/#v2:EYLgtghglgdgPgAQAwAIEEYB0ARKEDmMA9gM4AuUAxiZgMJEAmApgIIwQA2AniVCQNwBYAFAiEAZhQRg5AE4RKZNACYUAIQgkmIgN4iUBlAG0AykzIkASkwCOAVyiymDALJMwwJrJIBdfYYl1TSYACgxUBEj0JABKfwM9YUNklDIACz5MEzJZWHwUAF40KKQhJMMAX3iUasCne0dnKRkchSVwlGzcmHydFHxzfhRYKDIhquEJsUkEVWwvKAA3JpAgrV1q03MrWwcnV3dPbz9yg0D53OWGMOji8JiUVeBgsJK405RE5KmPwKJl2S5Zgoep7JodLp5T79QbDGCjcYiCZAA)):

```cs
using System.Diagnostics.CodeAnalysis;

public abstract class Base
{
    [SetsRequiredMembers]
    public Base(string @string)
    {
        this.String = @string;
    }
    
    public required abstract string String { get; init; }
}

public class Derived : Base
{
    [SetsRequiredMembers]
    public Derived(string @string) : base(@string)
    {
    }

    public override required string String { get; init; }
}
```

**Diagnostic Id**:

CS8618: Non-nullable property 'String' must contain a non-null value when exiting constructor. Consider declaring the property as nullable.

**Expected Behavior**:

No diagnostics, because the base class sets the value.

**Actual Behavior**:

A diagnotic on the constructor of `Derived`.

**Additional Information**

If I remove the `: Base` from `Derived` and make it an independent class, I get the same warning. This leads me to suspect this is a case of the nullable reference type analysis not being sophisticated enough to determine I have set `String` in the base constructor. I am aware that NRT analysis doesn't support every pattern under the sun and there are many ways you can 'trick' it. If this is one of those cases, I apologize - but I thought it was worth mentioning all the same.

**Comment by jaredpar** [comment]
@333fred for comment as I'm unsure what the correct behavior is here.

**Comment by 333fred** [comment]
I told them to open this bug. It doesn't seem like we're correctly handling the virtual property here.
